Design/SANS/NewISISDesign/Design_Document.md

```diff
@@ -0,0 +1,350 @@
+# Design Document for ISIS SANS Backend Rewrite
+
+This document defines the design  for a new ISIS SANS back-end as required by
+MantidProject issue [#15875](https://github.com/mantidproject/mantid/issues/15875).
+
+The intention of the issue is to replace the organically grown and now difficult to
+maintain back-end of the ISIS SANS reduction work-flow.
+
+### Definition of terms
+
+| Term  | Definition|
+|------|------------|
+|ADS   | Analysis Data Service|
+| PI   | Python Interface |
+
+
+## Motivation
+
+The current SANS reduction system makes use of the so-called `ReductionSingleton`,
+which essentially stores the configurational state either directly, in `ReductionStep`s
+or in an `ISISInstrument` object. The `ReductionStep`s are deeply coupled to the
+`ReductionSingleton`, which makes it hard to unit test them
+(and probably explains why such tests do not exist).
+In addition there is a strong coupling of this `ReductionSingleton` to the
+GUI logic, which essentially renders the ISIS SANS reduction a monolithic block.
+
+A more modern approach is to use Mantid's `WorkflowAlgorithm`s. A sequence of
+algorithms perform the reduction, where each algorithm gets the information
+required for the reduction at the time it executes. The `WorkflowAlgorithm` is
+used as a unit of clearly defined work and **not** as a container for
+state information. This mechanism avoids the deep coupling as each algorithm is
+agnostic about its calling environment and the information flow is into the algorithms
+(except for resulting output workspaces).
+This creates an ideal environment for unit testing and improved documentation.
+
+
+## Reducer 2.0
+
+As per requirement [R.2.4](User_Requirements.md#R.2.4) the reduction of a single
+run should be the limit case of the reduction of a batch file. Currently the
+implementations are slightly different. Unifying the two approaches is beneficial to future maintenance.
+
+### SANS State
+
+The current ISIS SANS system suffers from a state which is distributed and stored
+in many different places. The state for a the SANS reduction system should be
+ localized and complete, i.e. it defines the entire reduction.
+
+A design document for a `SANSState` approach is available
```

Comment:
I'm interested in this. I think I need to talk to you about this in person. I would also talk to @MikeHart85 who's been implementing a workflow-engine for device simulation. I don't think the two things necessarily overlap, but you might benefit from his experience in this area.